src/runtime/objmodel.cpp

```diff
@@ -1705,8 +1711,37 @@ void setattrInternal(Box* obj, const std::string& attr, Box* val, SetattrRewrite
         // We don't need to to the invalidation stuff in this case.
         return;
     } else {
+        // Finally, check __setattr__
+        if (obj->cls->tp_setattr) {
+            rewrite_args = NULL;
+            REWRITE_ABORTED("");
+
+            int rtn = obj->cls->tp_setattr(obj, const_cast<char*>(attr.c_str()), val);
+            if (rtn)
+                throwCAPIException();
+            return;
+        }
+        Box* setattr = typeLookup(obj->cls, setattr_str, NULL);
+        if (setattr) {
+            rewrite_args = NULL;
+            REWRITE_ABORTED("");
+
+            // if we're dealing with a BoxedWrapperDescriptor wrapping
+            // PyObject_GenericSetAttr, skip calling __setattr__, as
+            // that will just re-enter us.
```

Comment:
The problem arises when an extension exposes a type that has its `tp_setattro` set to `PyObject_GenericSetAttr`.  This is totally valid unfortunately, and `_functools.c` does it for the `partial` type.  Without the check, `socket_test.py` fails because of this.